### PR TITLE
Show the text entity tooltip information and update the text value wh…

### PR DESCRIPTION
**Pull request by dvdvideo1234**
…en longer than needed

### CODE REVIEW

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 1]
```diff
@@ -283,37 +283,44 @@ function TOOL:Think()
   self:UpdateGhostWirejoystick( self.GhostEntity, self:GetOwner() )
 end
 
+local function trimTextEntry(pnText, sConvar, nLen)
+  local sTxt = pnSelf:GetText():Trim()
+        sTxt = sTxt:gsub("%s+", "0"):gsub("%W", "0")
+  if(sTxt:len() > nLen) then
+    sTxt = sTxt:sub(1, nLen)
+    ChangeTooltip(pnSelf)
```
Comment (MattJeanes) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:291]: Where is this function defined?

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 2]
```diff
@@ -283,37 +283,44 @@ function TOOL:Think()
   self:UpdateGhostWirejoystick( self.GhostEntity, self:GetOwner() )
 end
 
+local function trimTextEntry(pnText, sConvar, nLen)
+  local sTxt = pnSelf:GetText():Trim()
```
Comment (MattJeanes) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:287]: `pnSelf` is passed into this function but named `pnText` - how does this work?

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua (hunk 1 above)
Comment (dvdvideo1234) [reply addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:291]: Here: https://wiki.facepunch.com/gmod/Global.ChangeTooltip

I am testing and going to update one more time soon

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua (hunk 2 above)
Comment (dvdvideo1234) [reply addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:287]: Thanks ... Copied from the control panel. Fixed .

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua (hunk 1 above)
Comment (dvdvideo1234) [reply addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:291]: The idea is to use the tooltip as a user hunt

**Comment by dvdvideo1234** [comment]
I will now add `wire_joystick_multi_` to be gathered from `TOOL.Mode` populated by the file name..

**Comment by dvdvideo1234** [comment]
Think of adding a new button.. One that when clicker restores the convars into the control panel

**Comment by dvdvideo1234** [comment]
@MattJeanes 
Done ;) now the tool name is taken from the file name which means no `wire_joystick_multi_` flying around

**Review by MattJeanes — CHANGES_REQUESTED** [review]
Mostly everything is working well, the text all looks good but there are 2 issues:

- The Joystick Configuration button in the menu no longer works
- When typing into a full text box it jumps back to the start and starts pushing the text off the end of the string, I've attached a gif below to show the weird behaviour. It should instead just prevent further typing

**Comment by dvdvideo1234** [comment]
@MattJeanes 

This is because of the `17` character limit of the UID to send the feedback to the user with the actual convar value. If you do not go over the `17` character limit will be like a regular text entity and it will not trim the data.

Please ty also with non alphanumeric characters, as I see the original author intended UID to be hesh-like value.

**Comment by dvdvideo1234** [comment]
Oh, and I also added presets as they were missing. Cheers !

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 3]
```diff
@@ -389,11 +389,19 @@ local function trimTextEntry(pnText, sConvar, nLen)
   RunConsoleCommand(sConvar, sNew) -- Send to tool script convar
 end
 
+local gtConvarList = TOOL:BuildConVarList()
+
 function TOOL.BuildCPanel(panel)
   panel:ClearControls()
+  local pnPresets = vgui.Create("ControlPresets", panel)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:396]: Added control presets with the default values

**Comment by MattJeanes** [comment]
@dvdvideo1234 yes please prevent further input if the box is at capacity, the current behaviour results in weirdly overwriting the end of the string with new characters at the beginning, thanks

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 4]
```diff
@@ -10,24 +10,24 @@ TOOL.Command    = nil
...
-  language.Add( "tool."..gsToolModeOP..".analog_con", "Analog input" )
-  language.Add( "tool."..gsToolModeOP..".config"    , "Click this button to open joystick configuration. You can also righ click on the world" )
-  language.Add( "tool."..gsToolModeOP..".config_con", "Joystick Configuration" )
+  language.Add( "tool."..gsToolModeOP..".name"           , "Joystick Multi Tool (Wire)" )
+  language.Add( "tool."..gsToolModeOP..".desc"           , "Spawns a Joystick Module interface chip for use with the wire system." )
+  language.Add( "tool."..gsToolModeOP..".0"              , "Primary: Create/Update Joystick   Secondary: Copy Settings    Reload: Link to pod" )
+  language.Add( "tool."..gsToolModeOP..".1"              , "Now select the pod to link to, or anything other than a pod to revert.")
+  language.Add( "tool."..gsToolModeOP..".uid"            , "Unique identifier. No spaces, alphanumeric, 17 charater limit!" )
+  language.Add( "tool."..gsToolModeOP..".uid_con"        , "UID")
+  language.Add( "tool."..gsToolModeOP..".description"    , "Write some input description here. Maximum 20 characters! For example `Steering`" )
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:19]: Yes... In fact this is hardcoded also along the file. At Least utilized language.Add

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 5]
```diff
@@ -388,3 +375,80 @@ if CLIENT and joystick then
...
+  local pnText, pnName = pnBase:TextEntry(language.GetPhrase(psName.."_con"))
+  pnText.OnChange = function(pnSelf)
+    local sTxt = pnSelf:GetText()
+    local sPat, sNew = tostring(sPattern or ""), sTxt:Trim()
+        sNew = (sPat == "") and sNew or sNew:gsub("["..sPat.."]", "X")
+    if(sTxt:len() > nLen) then sNew = sNew:sub(1, nLen) end
+    if(sNew != sTxt) then ChangeTooltip(pnSelf) end
+    RunConsoleCommand(pnConv, sNew)
+  end
+  pnText.AllowInput = function(pnSelf, chData)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:402]: Stops the cursor from typing over the limit

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 6]
```diff
@@ -388,3 +375,80 @@ if CLIENT and joystick then
...
+    return ((pnSelf:GetText():len() >= nLen) and true or false)
+  end
+  pnText.OnLoseFocus = function(pnSelf)
+    local sVar = GetConVar(pnConv):GetString()
+          sVar = sVar:gsub("jm_", "")
+    pnSelf:SetText(sVar)
+  end
+  pnText.OnEnter = function(pnSelf)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:410]: Uses the text entry value to generate random valid string

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 7]
```diff
@@ -388,3 +375,80 @@ if CLIENT and joystick then
...
+  panel:Button(language.GetPhrase("tool."..gsToolModeOP..".config_con"), "joyconfig")
+    :SetTooltip(language.GetPhrase("tool."..gsToolModeOP..".config"))
+  panel:ControlHelp("Joystick configuration should be run after placing a chip. "..
+    "In order to change an existing binding, there must be only one chip with its UID left.\nOne UID allows for one input.\n\n"..
+    "Multiple devices with the same UID will receive from the same input, but may have different max/min settings.")
+  for i = 1, 8 do
+    local ID, pItem = tostring(i)
+    pItem = panel:Help(language.GetPhrase("tool."..gsToolModeOP..".lcontr_con").." "..ID)
+    pItem:SetTooltip(language.GetPhrase("tool."..gsToolModeOP..".lcontr"))
+    setupTextEntry(panel, "uid"        , ID, "%s%W", 17)
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:445]: Shortened the function call

**Review by dvdvideo1234 — COMMENTED** [review]
@MattJeanes Done ;)

**Comment by dvdvideo1234** [comment]
Do I need the `OnLoseFocus` to display the old variable value. It will be replaced with the new one anyway but it displays the convar value that is present within when you lose focus updates the panel. It will show you what was its value before you replace it.

**Review by MattJeanes — COMMENTED** [review]
The text limits and joystick configuration button now work, although I'm a bit confused as to the random string feature you have added, how does this work ingame and what is the purpose of it?

**Comment by dvdvideo1234** [comment]
@MattJeanes

Instead of the user to fill-up the text entities, he just writes a positive number and hits enter for the text field to fill-up instead with random alphanumeric string. This is a bit easier than manually inserting the hash UID like `2304gh82084h234n`. It automatically filled for you. But If you insist I can remove it ;)

**Comment by MattJeanes** [comment]
Oh okay I see, everything works it is not made very clear how this works ingame, maybe can you change the tooltip or something to explain how this works for users?

File: addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua [hunk 8]
```diff
@@ -16,6 +16,7 @@ if CLIENT then
   language.Add( "tool."..gsToolModeOP..".1"              , "Now select the pod to link to, or anything other than a pod to revert.")
   language.Add( "tool."..gsToolModeOP..".uid"            , "Unique identifier. No spaces, alphanumeric, 17 charater limit!" )
   language.Add( "tool."..gsToolModeOP..".uid_con"        , "UID")
+  language.Add( "tool."..gsToolModeOP..".autofill"       , "Write a positive number here and hit ENTER to trigger random text autofill")
```
Comment (dvdvideo1234) [inline addons/Joystick/lua/weapons/gmod_tool/stools/wire_joystick_multi.lua:19]: When the user hovers on the `UID` text entry label.

**Comment by dvdvideo1234** [comment]
@MattJeanes

1. Is is worth it to remove also the comments in the file like `joywarn` as the don't apply
2. Is it worth to just pass `i` instead of  `local strI = tostring(i)`. Lua will concatenate it either way.

**Review by MattJeanes — COMMENTED** [review]
Everything seems to be working except updating existing joystick interface chips by clicking on them, but I am not sure if this is an existing issue or not

**Comment by dvdvideo1234** [comment]
@MattJeanes Hmm, you sure it updates perfectly for me... Let me check again ...

**Comment by dvdvideo1234** [comment]
@MattJeanes It works for me: https://www.youtube.com/watch?v=Iykv3ZNWjRw

**Comment by dvdvideo1234** [comment]
There is more work on the addon, for example `ENT:Setup` and `ENT:Update` are identical, but lets leave that for another PR ^_^

**Comment by dvdvideo1234** [comment]
@MattJeanes Reduces the Vector() creation and saved trace from overwriting via `GetNormalSpawn`

**Comment by MattJeanes** [comment]
Looks to be working good now, thank you for your contribution it is much appreciated!

**Review by MattJeanes — APPROVED** [review]

**Pull request merged**